core/src/main/java/uk/gov/justice/services/core/handler/HandlerMethod.java

```diff
@@ -52,20 +58,34 @@ public HandlerMethod(final Object object, final Method method, final Class<?> ex
                     format("Handles method must have exactly one parameter; found %d", parameterTypes.length));
         }
 
-        if (!isEnvelope(parameterTypes[0])) {
+        final Class<?> envelopeType = parameterTypes[0];
+
+        if (!Envelope.class.isAssignableFrom(envelopeType)) {
             throw new IllegalArgumentException(
-                    format("Handler methods must take an JsonEnvelope as the argument, not a %s", parameterTypes[0]));
+                    format("Handler methods must take an JsonEnvelope or Envelope<T> as the argument, not a %s", envelopeType));
+        }
+
+        if(!envelopeType.equals(JsonEnvelope.class)) {
```

Comment:
Strictly speaking, we should navigate up the class heirachy until we hit Envelope and then get its generic parameter type. But lets not do that right now.